---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
library(importbio)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# importbio

<!-- badges: start -->
<!-- badges: end -->

A package for importing common bioinformatic file formats into R as
[tibbles](https://r4ds.had.co.nz/tibbles.html).

## Setup

You can install and update importbio using the `remotes` package. Use the
`install_github()` function to install the package.

```r
install.packages("remotes")
remotes::install_github('davetang/importbio')
```

Use the `update_packages()` function to update the package. You can check the
version using the `packageVersion()` base function.

```r
remotes::update_packages("importbio")
packageVersion("importbio")
```

## Examples

Import a VCF file.

```{r importvcf}
importvcf(system.file("extdata", "Pfeiffer.vcf.gz", package = "importbio"))
```

Import a BED file.

```{r importbed, message=FALSE}
importbed(system.file("extdata", "refseq.bed.gz", package = "importbio"))
```

Import a GTF file.

```{r importgtf}
importgtf(system.file("extdata", "gencode.v38.annotation.sample.gtf.gz", package = "importbio"))
```

Import a GFF3 file.

```{r importgff}
importgff(system.file("extdata", "gencode.v38.annotation.sample.gff3.gz", package = "importbio"))
```

Import BLAST output.

```{r importblast}
importblast(system.file("extdata", "blastout.tsv.gz", package = "importbio"))
```

Import NetMHCIIpan output.

```{r importnetmhciipan}
importnetmhciipan(system.file("extdata", "netmhciipan.out.gz", package = "importbio"))
```

Import bam-readcount output.

```{r importbamreadcount}
importbamreadcount(system.file("extdata", "hg38.bwa.bamreadcount.tsv.gz", package = "importbio"))
```

## Session info

```{r session_info, echo = FALSE}
sessionInfo()
```

## Contact

Email me at davetang dot org (no spaces).
