---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
library(importbio)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# importbio

<!-- badges: start -->
<!-- badges: end -->

A package for importing common bioinformatic file formats into R as
[tibbles](https://r4ds.had.co.nz/tibbles.html).

## Installation

You can install importbio using the `remotes` package.

``` r
install.packages("remotes")
remotes::install_github('davetang/importbio')
```

## Examples

Import a VCF file.

```{r importvcf}
importvcf(system.file("extdata", "Pfeiffer.vcf.gz", package = "importbio"))
```

Import a BED file.

```{r importbed, message=FALSE}
importbed(system.file("extdata", "refseq.bed.gz", package = "importbio"))
```

Import a GTF file.

```{r importgtf}
importgtf(system.file("extdata", "gencode.v38.annotation.sample.gtf.gz", package = "importbio"))
```

Import a GFF3 file.

```{r importgff}
importgff(system.file("extdata", "gencode.v38.annotation.sample.gff3.gz", package = "importbio"))
```

Import BLAST output.

```{r importblast}
importblast(system.file("extdata", "blastout.tsv.gz", package = "importbio"))
```

Import NetMHCIIpan output. There will be a warning because NetMHCIIpan contains
free text at the end of the file; you can ignore this but please run
`problems()` to double-check.

```{r importnetmhciipan}
importnetmhciipan(system.file("extdata", "netmhciipan.out.gz", package = "importbio"))
```

## Session info

```{r session_info, echo = FALSE}
sessionInfo()
```

## Contact

Email me at davetang dot org (no spaces).
